src/metabase/driver/sql/query_processor.clj

```diff
@@ -606,11 +606,17 @@
   [driver [_ & args]]
   (apply hx/call :concat (mapv (partial ->honeysql driver) args)))
 
+;; this pseudo-expression returns only the substring function name, to simplify driver implementations
+(defmethod ->honeysql [:sql ::substring-name] [_ _] :substring)
+
 (defmethod ->honeysql [:sql :substring]
-  [driver [_ arg start length]]
-  (if length
-    (hx/call :substring (->honeysql driver arg) (->honeysql driver start) (->honeysql driver length))
-    (hx/call :substring (->honeysql driver arg) (->honeysql driver start))))
+  [driver [_ & [arg start & args]]]
+  ;; Standard SQL SUBSTRING uses a start index of 1, and many drivers return unexpected results
+  ;; when start=0, so to make MBQL substring work consistently across all DBs, we update the start
+  ;; index here if necessary.
+  (let [start    (if (= 0 start) 1 start)
```

Comment:
The same question as above, when is this called with a literal 0? I would consider that a bug and not try to fix it up here.

Reply:
Do you think this is better handled as middleware or somewhere else? Restricting the input in the UI would remove the need to normalize the index value at all.

Reply:
I think failing with a nice error message is enough. And I'm not sure the "nice" message is worth a lot of effort, hopefully the DBs report it clearly enough.

Reply:
The frontend has been updated with a much nicer error message.
<img width="536" alt="Screenshot 2023-02-07 at 18 16 18" src="https://user-images.githubusercontent.com/92878045/217569732-8ffdb62a-df43-4125-9fcf-9894eb05228e.png">